src/java/org/apache/cassandra/cql3/functions/masking/ColumnMask.java

```diff
@@ -0,0 +1,252 @@
+        return String.format("%s(%s)", function.name(), StringUtils.join(arguments, ", "));
+    }
+
+    public void appendCqlTo(CqlBuilder builder)
+    {
+        builder.append(" MASKED WITH ").append(toString());
+    }
+
+    /**
+     * @return {@code true} if we know that the current cluster supports masked columns, or {@code false} if it either
+     * doesn't support it due to the presence of not-upgrades nodes, or we don't know if such old nodes exist.
+     */
+    public static boolean clusterSupportsMaskedColumns()
+    {
+        if (!Gossiper.instance.isEnabled())
+            return false;
+
+        long timeout = DatabaseDescriptor.getWriteRpcTimeout(TimeUnit.MILLISECONDS);
+        CassandraVersion minVersion = Gossiper.instance.getMinVersion(timeout, TimeUnit.MILLISECONDS);
+        return minVersion != null && minVersion.familyLowerBound.get().compareTo(CassandraVersion.CASSANDRA_4_1) > 0;
+    }
+
+    /**
+     * A parsed but not prepared column mask.
+     */
+    public final static class Raw
+    {
+        private static final Joiner JOINER = Joiner.on(',');
+
+        public final FunctionName name;
+        public final List<Term.Raw> rawPartialArguments;
+
+        public Raw(FunctionName name, List<Term.Raw> rawPartialArguments)
+        {
+            this.name = name;
+            this.rawPartialArguments = rawPartialArguments;
+        }
+
+        public ColumnMask prepare(String keyspace, String table, ColumnIdentifier column, AbstractType<?> type)
+        {
+            ScalarFunction function = findMaskingFunction(keyspace, table, column, type);
+
+            List<ByteBuffer> partialArguments = preparePartialArguments(keyspace, function);
+
+            return new ColumnMask(function, partialArguments);
+        }
+
+        private ScalarFunction findMaskingFunction(String keyspace, String table, ColumnIdentifier column, AbstractType<?> type)
+        {
+            List<AssignmentTestable> args = new ArrayList<>(rawPartialArguments.size() + 1);
+            args.add(type);
+            args.addAll(rawPartialArguments);
+
+            Function function = FunctionResolver.get(keyspace, name, args, keyspace, table, type);
+
+            if (function == null)
+                throw invalidRequest("Unable to find masking function for %s, " +
+                                     "no declared function matches the signature %s",
+                                     column, this);
+
+            if (function.isAggregate())
+                throw invalidRequest("Aggregate function %s cannot be used for masking table columns", this);
+
+            if (!function.isNative())
+                throw invalidRequest("User defined function %s cannot be used for masking table columns", this);
+
+            if (!(function instanceof MaskingFunction))
+                throw invalidRequest("Not-masking function %s cannot be used for masking table columns", this);
+
+            if (!function.returnType().equals(type))
```

Comment:
Note to self: will this fail on reversed types?

Reply:
I think it won't fail. We have several tests that use masking with reversed types, but I'm adding some more explicit tests.

Nevertheless, this works because the masking functions added by CASSANDRA-17941 take care of returning a reversed type if their input is reversed. Those functions are the only ones that we allow here, so probably we are fine. However, we might need to revisit this when we add support for using UDFs as masking functions in CASSANDRA-18071.

Reply:
> take care of returning a reversed type if their input is reversed
^ That is what I was wondering thx